File: julia/2022/day_10.jl
const crt_width = 40
const crt_height = 6

function main()
    text = read("input/2022/input_10.txt", String)
    sprite_xs = [1]
    x = 1
    for line in eachsplit(text, "\n", keepempty=false)
        v, _... = match(r"noop|addx (-?\d+)", line)
        if !isnothing(v)
            # Take an extra cycle for the add operation
            push!(sprite_xs, x)
            x += parse(Int, v)
        end
        push!(sprite_xs, x)
    end
    total = sum(i * sprite_xs[i] for i in 20:40:220)
    println("Part 1: $total")
    println("Part 2:")
    for y = 0:crt_height-1
        for x = 0:crt_width-1
            cycle = y * crt_width + x + 1
            sprite_x = sprite_xs[cycle]
            # Sprite is 3 pixels wide
            if x in sprite_x-1:sprite_x+1
                print("█")
            else
                print(" ")
            end
        end
        println()
    end
end

main()

File: kotlin/y2022/Day10.kt
package y2022

import java.io.File

const val CRT_WIDTH = 40
const val CRT_HEIGHT = 6

fun main() {
    val text = File("input/2022/input_10.txt").readText().trim()
    val xs = mutableListOf<Int>()
    var spriteX = 1
    for (line in text.lines()) {
        xs.add(spriteX)
        Regex("""addx (-?\d+)""").find(line)?.let {
            xs.add(spriteX)
            spriteX += it.destructured.component1().toInt()
        }
    }
    val signalSum = (20..220 step 40).sumOf { it * xs[it - 1] }
    println("Part 1: $signalSum")
    println("Part 2:")
    for (y in 0..<CRT_HEIGHT) {
        for (x in 0..<CRT_WIDTH) {
            spriteX = xs[y * CRT_WIDTH + x]
            print(if (x in (spriteX-1)..(spriteX+1)) { "█" } else { " " })
        }
        println()
    }
}
